### ISSUE TITLE
`needless_question_mark` lint not applied to async fn

### ISSUE THREAD

**Issue by Swatinem** [rust-lang/rust-clippy#8277]
### Summary

It seems like the `needless_question_mark` is not correctly applied to `async fn`, while it is applied to normal functions (obviously), and also async blocks correctly. See the reproduction below.

### Lint Name

needless_question_mark

### Reproducer

I tried this code:

```rust
pub fn normal_fn() -> Result<(), ()> {
    let res = Err(());
    Ok(res?) // warns
}
pub async fn async_fn() -> Result<(), ()> {
    let res = Err(());
    Ok(res?) // does not warn
}
pub fn async_block() -> impl std::future::Future<Output = Result<(), ()>> {
    async move {
        let res = Err(());
        Ok(res?) // warns
    }
}

```

I expected to see this happen:

All three variants should trigger a `needless_question_mark` lint.

Instead, this happened:

Only the `normal_fn` and `async_block` variants are triggering the lint.

```
warning: question mark operator is useless here
  --> src/main.rs:47:5
   |
47 |     Ok(res?) // warns
   |     ^^^^^^^^ help: try: `res`
   |
   = note: `#[warn(clippy::needless_question_mark)]` implied by `#[warn(clippy::all)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_question_mark

warning: question mark operator is useless here
  --> src/main.rs:56:9
   |
56 |         Ok(res?) // warns
   |         ^^^^^^^^ help: try: `res`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_question_mark
```

### Version

```text
rustc 1.58.0 (02072b482 2022-01-11)
binary: rustc
commit-hash: 02072b482a8b5357f7fb5e5637444ae30e423c40
commit-date: 2022-01-11
host: x86_64-apple-darwin
release: 1.58.0
LLVM version: 13.0.0

and

rustc 1.60.0-nightly (22e491ac7 2022-01-13)
binary: rustc
commit-hash: 22e491ac7ed454d34669151a8b6464cb643c9b41
commit-date: 2022-01-13
host: x86_64-apple-darwin
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by loyd** [comment]
It seems that `question_mark` also works incorrectly.
